### ISSUE TITLE
Support assertThrows() with suspending functions in Kotlin

### ISSUE THREAD

**Issue by DerkSchooltink** [junit-team/junit5#1851]
Often when I write a unit test that has to assert that a certain exception is thrown I use the `assertThrows()` method. But I find myself regularly having to test suspending functions, so I have to wrap my function with a `runBlocking()`. This is a lot of repeating code if your entire codebase is heavily based on coroutines. 

### Proposed fix

I couldn't possibly have made this issue without also thinking about a solution.

Effectively, this is what I came up with:

```kotlin
private inline fun <reified T : Throwable> assertThrows(
        noinline executable: suspend () -> Unit
    ) = assertThrows(T::class.java) {
        runBlocking {
            executable()
        }
    }
```

This way you can pass suspending functions as parameter `executable`, which will be invoked in a `runBlocking()` to make sure the suspending function joins before continuing.

### tl;dr

I think it would be a good addition to JUnit Jupiter to have functions that allow easy testing of suspending functions.

**Comment by marcphilipp** [comment]
Thanks for creating the issue! While I'm not familiar with Kotlin's coroutines and suspending functions, but I can see how that would make sense.

@DerkSchooltink Would you be interested in submitting a PR?

**Comment by DerkSchooltink** [comment]
> Thanks for creating the issue! While I'm not familiar with Kotlin's coroutines and suspending functions, but I can see how that would make sense.
> 
> @DerkSchooltink Would you be interested in submitting a PR?

**Comment by mkobit** [comment]
Probably related to https://youtrack.jetbrains.com/issue/KT-22228

**Comment by marcphilipp** [comment]
Reopening due to https://github.com/junit-team/junit5/pull/2042#discussion_r400912817

**Comment by marcphilipp** [comment]
Resolves in `master` in [c5d658c](https://github.com/junit-team/junit5/commit/c5d658c5aa8e5550410e9691d90e3216fe9a55e4).

**Comment by binkley** [comment]
Sorry to comment on a long-closed issue.  I tried the "simplest thing".  If the unit test function is not itself suspending, how does it test a suspending function?
```kotlin
internal class ScratchTest {
    @Test
    fun `should test suspending functions`() {
        assertThrows<IllegalStateException> { failing() }
    }
}

private suspend fun failing() {
    delay(1L) // pretend
    error("Fail on purpose")
}
```
The compiler rightfully complains that "failing()" cannot be called from "assertThrows" as there is no suspension context.  What is the scope and context for test functions?  (https://kotlinlang.org/docs/reference/coroutines/basics.html)

**Comment by ashdavies** [comment]
Hi @binkley, the design of `assertThrows` is to allow you to build your own `CoroutineContext` as to not interfere with your test setup, as such I would recommend using the Kotlin Coroutines Test dependency (https://github.com/Kotlin/kotlinx.coroutines/tree/master/kotlinx-coroutines-test) to run suspended functions under test.

```
internal class ScratchTest {

    @Test
    fun `should test suspending functions`() **= runBlockingTest** {
        assertThrows<IllegalStateException> { failing() }
    }
}

private suspend fun failing() {
    delay(1L) // pretend
    error("Fail on purpose")
}
```

**Comment by williamboman-pp** [comment]
Hey! This does not seem to work with the `assertDoesNotThrow` function, only wtih `assertThrows`

**Comment by sbrannen** [comment]
@williamboman-pp, this issue was closed over a year ago.

Please open a new issue to address `assertDoesNotThrow`.

Thanks

**Comment by CyxouD** [comment]
@ashdavies  Doesn't work with Kotlin 1.8.0 and JUnit 4.13.2 (see screenshot). Furthermore, `runBlockingTest` is deprecated
<img width="1016" alt="Знімок екрана 2023-09-26 о 11 01 06" src="https://github.com/junit-team/junit5/assets/17162008/88eb21ba-5091-4432-a93a-21f715fe2ef8">
